csharp/src/Ice/Communicator-OutgoingConnectionFactory.cs

```diff
@@ -0,0 +1,210 @@
+                    if (_outgoingConnections.TryGetValue((endpoint, label),
+                                                         out List<Connection>? connections) &&
+                        connections.FirstOrDefault(
+                            connection => connection.IsActive && connection.CanTrust(preferNonSecure))
+                        is Connection connection)
+                    {
+                        return connection;
+                    }
+                }
+                return null;
+            }
+        }
+
+        internal IEnumerable<Endpoint> OrderEndpointsByTransportFailures(IEnumerable<Endpoint> endpoints)
+        {
+            // Purge expired transport failures
+            DateTime expirationDate = DateTime.Now - TimeSpan.FromSeconds(5);
+            foreach ((Endpoint endpoint, DateTime date) in _transportFailures)
+            {
+                if (date <= expirationDate)
+                {
+                    _ = ((ICollection<KeyValuePair<Endpoint, DateTime>>)_transportFailures).Remove(
+                        new KeyValuePair<Endpoint, DateTime>(endpoint, date));
+                }
+            }
+            return endpoints.OrderBy(
+                endpoint => _transportFailures.TryGetValue(endpoint, out DateTime value) ? value : default);
+        }
+
+        internal void Remove(Connection connection, object? label)
+        {
+            lock (_mutex)
+            {
+                List<Connection> list = _outgoingConnections[(connection.Endpoint, label)];
+                list.Remove(connection);
+                if (list.Count == 0)
+                {
+                    _outgoingConnections.Remove((connection.Endpoint, label));
+                }
+            }
+        }
+
+        internal void SetRouterInfo(RouterInfo routerInfo)
+        {
+            ObjectAdapter? adapter = routerInfo.Adapter;
+            IReadOnlyList<Endpoint> endpoints = routerInfo.GetClientEndpoints(); // can make a synchronous remote call
+
+            // Search for connections to the router's client proxy endpoints, and update the object adapter for
+            // such connections, so that callbacks from the router can be received over such connections.
+            lock (_mutex)
+            {
+                if (_disposeTask != null)
+                {
+                    throw new CommunicatorDisposedException();
+                }
+
+                foreach (Endpoint endpoint in endpoints)
+                {
+                    foreach (ICollection<Connection> connections in _outgoingConnections.Values)
+                    {
+                        foreach (Connection connection in connections)
+                        {
+                            connection.Adapter = adapter;
+                        }
+                    }
+                }
+            }
+        }
+
+        private class ConnectionMapKeyComparer : EqualityComparer<(Endpoint Endpoint, object? Label)>
```

Comment:
I suggest to rename this class EndpointComparer (since it compares endpoints), make it abstract, and change its static field into:
```
internal static EndpointComparer Equivalent { get; } = new EquivalentEndpointComparer();
```

And then derive EquivalentEndpointComparer from EndpointComparer.